### ISSUE TITLE
Configuration level disabling algorithms in default and legacy provider

### ISSUE THREAD

**Issue by beldmit** [openssl/openssl#25406]
I would like to propose a feature: explicit allow/blocklist of algorithms in default and legacy provider.

Rationale:
- if we include PQ algos in the default provider then, having an implementation in an alternative provider can increase some flexibility without dealing with propqueries
- It may make sense to disable, e.g. DSA by default instead of building with no-dsa
- For legacy provider it may make sense enabling only strictly necessary algorithms

**Comment by paulidale** [comment]
Could you please elucidate here.  It's not entirely clear what is being asked for.
An example might be best…

**Comment by beldmit** [comment]
@paulidale I want syntax like

```
[provider_sect]
default = default_sect
legacy = legacy_sect

[default_sect]
activate = 1
#enable all algs except DSA
algs_disable = DSA 

[legacy_sect]
activate = 1
#enable only RC4
algs_enable = RC4
```

**Comment by nhorman** [comment]
What does the above imply in terms of a default state?  i.e.:
1) if alg_enable is specified, are all other algorithms disabled?
2) if alg_disable is specified, are all other algs enabled?
3) If both are specified, is that an error state, or does the last element take precedence?

Not disagreeing with the idea mind you, just trying to get clarity on how it might work.

As a side note, some method for expressing algorithm inclusion in a provider might be helpful for things like deprecations at run time (i.e. if an update moves an alg from the default to the legacy provider, this might be a nifty way to identify where that provider has moved to at run time without the need to expressly update an application to pull everything from a provider in

**Comment by beldmit** [comment]
@nhorman yes, I have in mind that syntax. Or probably syntax like `algs=-DSA` or `algs=RC4` also makes sense.

Speaking about depreciation at run time, I'd better thought about crypto policies present in Red Hat patches (and I have heard rumours that Ubuntu is working on smth similar)

**Comment by paulidale** [comment]
Maybe more akin to:
```
algs += DSA      Include DSA
algs -= RC2      Exclude RC2
algs  = RC4      Only include RC4
````

We'll want wilcards of some kind too:
```
algs -= *CBC     Omit all CBC mode ciphers
algs  = *256*    Only include 256 bit things
```

**Comment by t8m** [comment]
> We'll want wilcards of some kind too:
> 
> ```
> algs -= *CBC     Omit all CBC mode ciphers
> algs  = *256*    Only include 256 bit things
> ```

The question is whether the matching should be against the algorithm names only, or whether it is supposed to be against some _provider-specific_ knowledge about the algorithms it implements. I would be very much against the later as that would add additional level of complexity for provider authors and their users.

If the matching is supposed to be against the algorithm names, it can be implemented solely at the libcrypto level.

**Comment by beldmit** [comment]
@paulidale I don't understand what is the semantics of `algs += DSA` syntax you propose - add to list of what?

@t8m not sure about libcrypto level, I have per-provider setup in my mind. But yes, name-only filtering.

**Comment by t8m** [comment]
> @t8m not sure about libcrypto level, I have per-provider setup in my mind. But yes, name-only filtering.

You misunderstood me - By saying libcrypto level I meant there would be no changes in providers' code. But the configuration could be of course per-provider.

**Comment by beldmit** [comment]
> > @t8m not sure about libcrypto level, I have per-provider setup in my mind. But yes, name-only filtering.
> 
> You misunderstood me - By saying libcrypto level I meant there would be no changes in providers' code. But the configuration could be of course per-provider.

So you mean that the directive `algs` will be per-provider but processed on the same level as `activate`, e.g. filling the list of algs? Yes, it would work for me.
